mail/protocols/imap/src/test/java/com/fsck/k9/mail/store/imap/ImapResponseParserTest.kt

```diff
@@ -0,0 +1,579 @@
+
+    @Test
+    fun `readResponse() with partly consuming callback that throws should read all data and throw`() {
+        val parser = createParserWithResponses("* {4}\r\ntest")
+        val callback = TestImapResponseCallback.readBytesAndThrow(2)
+
+        assertThat {
+            parser.readResponse(callback)
+        }.isFailure()
+            .isInstanceOf(ImapResponseParserException::class)
+            .all {
+                hasMessage("readResponse(): Exception in callback method")
+                cause().isNotNull().isInstanceOf(ImapResponseParserTestException::class)
+            }
+        assertThatAllInputWasConsumed()
+    }
+
+    @Test
+    fun `readResponse() with callback that throws repeatedly should consume all input and throw first exception`() {
+        val parser = createParserWithResponses("* {3}\r\none {3}\r\ntwo")
+        val callback = TestImapResponseCallback.readBytesAndThrow(3)
+
+        assertThat {
+            parser.readResponse(callback)
+        }.isFailure()
+            .isInstanceOf(ImapResponseParserException::class)
+            .all {
+                hasMessage("readResponse(): Exception in callback method")
+                cause().isNotNull().isInstanceOf(ImapResponseParserTestException::class)
+                    .prop(ImapResponseParserTestException::instanceNumber).isEqualTo(0)
+            }
+        assertThatAllInputWasConsumed()
+    }
+
+    @Test
+    fun `readResponse() with callback not consuming the entire literal should skip the rest of the literal`() {
+        val parser = createParserWithResponses("* {3}\r\none two")
+        val callback = TestImapResponseCallback.readBytesAndReturn(2, "replacement value")
+
+        val response = parser.readResponse(callback)
+
+        assertThat(response).containsExactly("replacement value", "two")
+        assertThatAllInputWasConsumed()
+    }
+
+    @Test
+    fun `readResponse() with callback not consuming and throwing should read response and throw`() {
+        val parser = createParserWithResponses("* {4}\r\ntest")
+        val callback = TestImapResponseCallback.readBytesAndThrow(0)
+
+        assertThat {
+            parser.readResponse(callback)
+        }.isFailure()
+            .isInstanceOf(ImapResponseParserException::class)
+            .hasMessage("readResponse(): Exception in callback method")
+        assertThatAllInputWasConsumed()
+    }
+
+    @Test
+    fun `readResponse() with callback throwing IOException should re-throw that exception`() {
+        val parser = createParserWithResponses("* {4}\r\ntest")
+        val exception = IOException()
+        val callback = ImapResponseCallback { _, _ -> throw exception }
+
+        assertThat {
+            parser.readResponse(callback)
+        }.isFailure()
+            .isEqualTo(exception)
+    }
+
+    @Test
+    fun `readResponse() with quoted string containing an escaped quote character`() {
+        val parser = createParserWithResponses("""* "qu\"oted"""")
```

Comment:
Just noticed that this changed to a raw string, which is better readable, while for example line 424 changed to single quotes with escape characters.

Reply:
In 22e499301db6ba54147a3c55169ab3d9ebe635d8 I converted a couple of other string to raw strings.

This instance in particular was a hard call. Using a raw string makes it easier to see the non-Java escaping. However, it also obscures the fact that the last character in the string is a quote character.

Reply:
Fine with me